pkg/server/stream/v3/subscription.go

```diff
@@ -0,0 +1,176 @@
+	// This list might be non-empty even when set as wildcard.
+	subscribedResourceNames map[string]struct{}
+
+	// returnedResources contains the resources acknowledged by the client and the acknowledged versions.
+	returnedResources map[string]string
+}
+
+// newSubscription initializes a subscription state.
+func newSubscription(wildcard bool, initialResourceVersions map[string]string) Subscription {
+	state := Subscription{
+		wildcard:                wildcard,
+		allowLegacyWildcard:     wildcard,
+		subscribedResourceNames: map[string]struct{}{},
+		returnedResources:       initialResourceVersions,
+	}
+
+	if initialResourceVersions == nil {
+		state.returnedResources = make(map[string]string)
+	}
+
+	return state
+}
+
+func NewSotwSubscription(subscribed []string) Subscription {
+	sub := newSubscription(len(subscribed) == 0, nil)
+	sub.SetResourceSubscription(subscribed)
+	return sub
+}
+
+// SetResourceSubscription updates the subscribed resources (including the wildcard state)
+// based on the full state of subscribed resources provided in the request
+// Used in sotw subscriptions
+// Behavior is based on
+// https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return
+func (s *Subscription) SetResourceSubscription(subscribed []string) {
+	if s.allowLegacyWildcard {
+		if len(subscribed) == 0 {
+			// We were wildcard based on legacy behavior and still don't request any resource
+			// The watch remains wildcard
+			return
+		}
+
+		// A resource was provided (might be an explicit wildcard)
+		// Documentation states that we should no longer allow to fallback to the previous case
+		// and no longer setting wildcard would no longer subscribe to anything
+		s.allowLegacyWildcard = false
+	}
+
+	subscribedResources := make(map[string]struct{}, len(subscribed))
+	explicitWildcardSet := false
+	for _, resource := range subscribed {
+		if resource == explicitWildcard {
+			explicitWildcardSet = true
+		} else {
+			subscribedResources[resource] = struct{}{}
+		}
+	}
+
+	// Explicit subscription to wildcard as we are not in legacy wildcard behavior
+	s.wildcard = explicitWildcardSet
+	s.subscribedResourceNames = subscribedResources
+}
+
+func NewDeltaSubscription(subscribed, unsubscribed []string, initialResourceVersions map[string]string) Subscription {
+	sub := newSubscription(len(subscribed) == 0, initialResourceVersions)
+	sub.UpdateResourceSubscriptions(subscribed, unsubscribed)
+	return sub
+}
+
+// UpdateResourceSubscriptions updates the subscribed resources (including the wildcard state)
+// based on newly subscribed or unsubscribed resources
+// Used in delta subscriptions
+func (s *Subscription) UpdateResourceSubscriptions(subscribed, unsubscribed []string) {
+	// Handles legacy wildcard behavior first to exit if we are still in this behavior
+	if s.allowLegacyWildcard {
+		// The protocol (as of v1.29.0) only references subscribed as triggering
```

Comment:
nit: the version can be updated to v1.36.0

Reply:
I'll update in the next PR if possible, to avoid rebuilding/rebasing